x-pack/plugins/security_solution/public/timelines/components/edit_data_provider/index.tsx

```diff
@@ -277,3 +303,127 @@ export const StatefulEditDataProvider = React.memo<Props>(
 );
 
 StatefulEditDataProvider.displayName = 'StatefulEditDataProvider';
+
+function isStringOrNumberArray(
+  val: string | number | Array<string | number>
+): val is Array<string | number> {
+  return Array.isArray(val) && (typeof val[0] === 'string' || typeof val[0] === 'number');
+}
+
+const isValueFieldInvalid = (type: DataProviderType, value: string | number) =>
+  type !== DataProviderType.template &&
+  (startsWith('{', sanatizeValue(value)) || endsWith('}', sanatizeValue(value)));
+
+interface ControlledDataProviderInput {
+  disableButtonCallback: (disableButton: boolean) => void;
+  onChangeCallback: (value: string | number | string[]) => void;
+  type: DataProviderType;
+  value: string | number | Array<string | number>;
+}
+
+const getDefaultValue = (value: string | number | Array<string | number>): string | number => {
+  if (isStringOrNumberArray(value)) {
+    return value[0];
+  } else return value;
+};
+
+const ControlledDefaultInput = ({
+  value,
+  type,
+  disableButtonCallback,
+  onChangeCallback,
+}: ControlledDataProviderInput) => {
+  const [primitiveValue, setPrimitiveValue] = useState<string | number>(getDefaultValue(value));
+
+  const isInvalid = useMemo(
+    () => isValueFieldInvalid(type, primitiveValue),
+    [type, primitiveValue]
+  );
+
+  useEffect(() => {
+    onChangeCallback(primitiveValue);
+    disableButtonCallback(isInvalid);
+  }, [primitiveValue, isInvalid, onChangeCallback, disableButtonCallback]);
+
+  const onValueChange = useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
+    setPrimitiveValue(e.target.value);
+  }, []);
+
+  return (
+    <EuiFieldText
+      className={VALUE_INPUT_CLASS_NAME}
+      onChange={onValueChange}
+      placeholder={i18n.VALUE}
+      value={sanatizeValue(primitiveValue)}
+      isInvalid={isInvalid}
+    />
+  );
+};
+
+const convertValuesToComboboxValueArray = (
```

Comment:
Consider extracting this function, and similar ones below like `convertComboboxValuesToStringArray` out of this component, because these pure functions can be unit tested independently from the component.

Reply:
done!